libretroshare/src/services/p3msgservice.cc

```diff
@@ -1192,14 +1211,125 @@ bool 	p3MsgService::MessageSend(MessageInfo &info)
+        const RsGxsId from,
+        const std::string& subject,
+        const std::string& body,
+        const std::set<RsGxsId>& to,
+        const std::set<RsGxsId>& cc,
+        const std::set<RsGxsId>& bcc,
+        const std::vector<FileInfo>& attachments,
+        std::set<RsMailTrackId>& trackingIds,
+        std::string& errorMsg )
+{
+	errorMsg.clear();
+	const std::string fname = __PRETTY_FUNCTION__;
+	auto pCheck = [&](bool test, const std::string& errMsg)
+	{
+		if(!test)
+		{
+			errorMsg = errMsg;
+			RsErr() << fname << " " << errMsg << std::endl;
+		}
+		return test;
+	};
+
+	if(!pCheck(!from.isNull(), "from can't be null")) return false;
+	if(!pCheck( rsIdentity->isOwnId(from),
+	            "from must be own identity") ) return false;
+	if(!pCheck(!(to.empty() && cc.empty() && bcc.empty()),
+	            "You must specify at least one recipient" )) return false;
+
+	auto dstCheck =
+	        [&](const std::set<RsGxsId>& dstSet, const std::string& setName)
+	{
+		for(const RsGxsId& dst: dstSet)
+		{
+			if(dst.isNull())
+			{
+				errorMsg = setName + " contains a null recipient";
+				RsErr() << fname << " " << errorMsg << std::endl;
+				return false;
+			}
+
+			if(!rsIdentity->isKnownId(dst))
+			{
+				rsIdentity->requestIdentity(dst);
+				errorMsg = setName + " contains an unknown recipient: " +
+				        dst.toStdString();
+				RsErr() << fname << " " << errorMsg << std::endl;
+				return false;
+			}
+		}
+		return true;
+	};
+
+	if(!dstCheck(to, "to"))   return false;
+	if(!dstCheck(cc, "cc"))   return false;
+	if(!dstCheck(bcc, "bcc")) return false;
+
+	MessageInfo msgInfo;
+
+	msgInfo.rsgxsid_srcId = from;
+	msgInfo.title = subject;
+	msgInfo.msg = body;
+	msgInfo.rsgxsid_msgto = to;
+	msgInfo.rsgxsid_msgcc = cc;
+	msgInfo.rsgxsid_msgbcc = bcc;
+	std::copy( attachments.begin(), attachments.end(),
+	           std::back_inserter(msgInfo.files) );
+
+	uint32_t ret = 0;
+	using Evt_t = RsMailStatusEvent;
+	std::shared_ptr<Evt_t> pEvent(new Evt_t());
+
+	auto pSend = [&](const std::set<RsGxsId>& sDest)
+	{
+		for(const RsGxsId& dst : sDest)
+		{
+			RsMsgItem* msgItem = initMIRsMsg(msgInfo, dst);
+			if(!msgItem)
+			{
+				errorMsg += " initMIRsMsg from: " + from.toStdString()
+				        + " dst: " + dst.toStdString() + " subject: " + subject
+				        + " returned nullptr!\n";
+				RsErr() << fname << errorMsg;
+				continue;
+			}
+
+			uint32_t msgId = sendDistantMessage(msgItem, from);
+			// ensure we don't use that ptr again without noticing
+			msgItem = nullptr;
+
+			if(!msgId)
+			{
+				errorMsg += " sendDistantMessage from: " + from.toStdString()
+				        + " dst: " + dst.toStdString() + " subject: " + subject
+				        + " returned 0!\n";
+				RsErr() << fname << errorMsg;
+				continue;
+			}
+
+			const RsMailMessageId mailId = std::to_string(msgId);
+			pEvent->mChangedMsgIds.insert(mailId);
+			trackingIds.insert(RsMailTrackId(mailId, dst));
+			++ret;
+		}
+	};
+
+	pSend(to);
```

Comment:
It's surprising that the msg structure is not modified when you send in bcc. Normally the list of recipients should be cleared right?

Reply:
I expected that too, but didn't see that in the old code, maybe retroshare-gui does that?

Reply:
I have double checked the code and the BCC filtering is done at lines 1898-1904